solutions/testing/middleware.ts

```diff
@@ -0,0 +1,27 @@
+import { NextRequest, NextResponse } from 'next/server'
+import db from './lib/db'
+
+export const config = {
+  matcher: ['/', '/signup'],
+}
+
+export default async function middleware(req: NextRequest) {
+  const url = req.nextUrl
+  const user = await db.getUserFromReq(req)
+  let redirectTo
+
+  if (!user && url.pathname === '/') {
+    redirectTo = '/signup'
+  }
+  if (user && url.pathname === '/signup') {
+    redirectTo = '/'
+  }
+  if (redirectTo) {
+    req.nextUrl.pathname = redirectTo
+    const res = NextResponse.redirect(req.nextUrl)
+    // Don't cache the redirect so that we can redirect from '/signup'
+    // to '/' in the browser with a client side transition
+    res?.headers.set('x-middleware-cache', 'no-cache')
```

Comment:
Cool, but feels weird that users have to do this manually, if they don't know they will lose optimizations, should we report that?

Reply:
Good point. Added to our Notion on documentation improvements, I'll share that with the docs team.